core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala

```diff
@@ -269,59 +257,47 @@ class LoadBalancerService(
+        val managed = Math.max(1, invokers.length - numBlackbox(invokers.length))
+        invokers.take(managed)
     }
 
     /** Determine which invoker this activation should go to. Due to dynamic conditions, it may return no invoker. */
-    private def chooseInvoker(action: ExecutableWhiskAction, msg: ActivationMessage): Future[String] = {
-        val invokers = if (action.exec.pull) blackboxInvokers else managedInvokers
-        val hash = hashSubjectAction(msg)
-
-        invokers.flatMap { invokers =>
-            LoadBalancerService.schedule(
-                invokers,
-                activationByInvoker.mapValues(_.size),
-                config.loadbalancerInvokerBusyThreshold,
-                hash) match {
-                    case Some(invoker) => Future.successful(invoker)
-                    case None =>
-                        logging.error(this, s"all invokers down")(TransactionId.invokerHealth)
-                        Future.failed(new LoadBalancerException("no invokers available"))
-                }
+    private def chooseInvoker(user: Identity, action: ExecutableWhiskAction): Future[InstanceId] = {
+        val hash = generateHash(user.namespace, action)
+
+        allInvokers.flatMap { invokers =>
+            val invokersToUse = if (action.exec.pull) blackboxInvokers(invokers) else managedInvokers(invokers)
+            val invokersWithUsage = invokersToUse.view.map {
+                // Using a view defers the comparably expensive lookup to actual access of the element
+                case (instance, state) => (instance, state, activationByInvoker.get(instance).map(_.size).getOrElse(0))
+            }
+
+            LoadBalancerService.schedule(invokersWithUsage, config.loadbalancerInvokerBusyThreshold, hash) match {
+                case Some(invoker) => Future.successful(invoker)
+                case None =>
+                    logging.error(this, s"all invokers down")(TransactionId.invokerHealth)
+                    Future.failed(new LoadBalancerException("no invokers available"))
+            }
         }
     }
 
-    /*
-     * The path contains more than the action per se but seems sufficiently
-     * isomorphic as the other parts are constant.  Extracting just the
-     * action out specifically will involve some hairy regex's that the
-     * Invoker is currently using and which is better avoid if/until
-     * these are moved to some common place (like a subclass of Message?)
-     */
-    private def hashSubjectAction(msg: ActivationMessage): Int = {
-        val subject = msg.user.subject.asString
-        val path = msg.action.toString
-        (subject.hashCode() ^ path.hashCode()).abs
+    /** Generates a hash based on the string representation of namespace and action */
+    private def generateHash(namespace: EntityName, action: ExecutableWhiskAction): Int = {
+        (namespace.asString.hashCode() ^ action.fullyQualifiedName(false).asString.hashCode()).abs
```

Comment:
The scheduling in the Invoker happens based on the namespace, so it only makes sense to schedule based on namespaces here.